In [50]:
# import numpy and pandas libraries
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.impute import SimpleImputer
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from scipy import stats
np.random.seed(1)

In [51]:
#Importing the Csv Data files of test and train

X_test= pd.read_csv("C:/Users/ssand/OneDrive/Desktop/pre/X_test.csv")
X_train= pd.read_csv("C:/Users/ssand/OneDrive/Desktop/pre/X_train.csv")
y_test= pd.read_csv("C:/Users/ssand/OneDrive/Desktop/pre/y_test.csv")
y_train= pd.read_csv("C:/Users/ssand/OneDrive/Desktop/pre/y_train.csv")

logistic regression

In [52]:
score_measure = "precision"
kfolds = 5

param_grid = [{'penalty':['l1','l2']}, 
              {'C':[1, 10, 100, 1000]}]

lr = LogisticRegression()
grid_search = GridSearchCV(estimator =lr, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestRecallTree = grid_search.best_estimator_

Fitting 5 folds for each of 6 candidates, totalling 30 fits
The best precision score is 0.7
... with parameters: {'C': 1000}


C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\model_selection\_validation.py:378: FitFailedWarning: 
5 fits failed out of a total of 30.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
5 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\model_selection\_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\linear_model\_logistic.py", line 1091, in fit
    solver = _check_solver(self.solver, self.penalty, self.dual)
  File "C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\linear_mode

We done the logistic regression with grid search the precision score is 0.7 and the precision for this seen better and going further we will the precision score for other models

In [53]:
score_measure = "precision"
kfolds = 5

param_grid = [
              {'C':[1, 10, 100, 1000]}]


gs = LogisticRegression()
rand_search = RandomizedSearchCV(estimator = gs, param_distributions=param_grid, cv=kfolds, n_iter=300,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)
_ = rand_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

bestRecall= rand_search.best_estimator_

Fitting 5 folds for each of 4 candidates, totalling 20 fits
The best precision score is 0.7
... with parameters: {'C': 1000}


C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\model_selection\_search.py:306: UserWarning: The total space of parameters 4 is smaller than n_iter=300. Running 4 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(
C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\utils\validation.py:1111: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


The logistic regression with random search the precision value is 0.7 which is better and we got the same for both grid and random search. 

svm linear model

In [54]:
score_measure = "precision"
kfolds = 5

param_grid = [  {'C':[1, 10, 100, 1000]}]

sm =SVC()
grid_search = GridSearchCV(estimator =sm, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestRecallTree = grid_search.best_estimator_

Fitting 5 folds for each of 4 candidates, totalling 20 fits
The best precision score is 0.6
... with parameters: {'C': 1}


C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\utils\validation.py:1111: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


The grid search using SVM linear model the prescision score is 0.6 which is less than the logistic model

In [55]:
score_measure = "precision"
kfolds = 5

param_grid = [
              {'C':[1, 10, 100, 1000]}]


gs = SVC()
rand_search = RandomizedSearchCV(estimator = gs, param_distributions=param_grid, cv=kfolds, n_iter=300,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)
_ = rand_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

bestRecall= rand_search.best_estimator_

Fitting 5 folds for each of 4 candidates, totalling 20 fits
The best precision score is 0.6
... with parameters: {'C': 1}


C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\model_selection\_search.py:306: UserWarning: The total space of parameters 4 is smaller than n_iter=300. Running 4 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(
C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\utils\validation.py:1111: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


The randomsearch using SVM linear model the precision is 0.6 abd its same for grid and random search

SVM with rbf kernel

In [56]:
score_measure = "precision"
kfolds = 5

param_grid = [
              {'C':[1, 10, 100, 1000]}]

Sk = SVC()
grid_search = GridSearchCV(estimator =Sk, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestRecallTree = grid_search.best_estimator_

C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\utils\validation.py:1111: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Fitting 5 folds for each of 4 candidates, totalling 20 fits
The best precision score is 0.6
... with parameters: {'C': 1}


We done SVM with rbf kernel the precision score is 0.6 which is same as Svm with linear model for both grid and random search

In [57]:
score_measure = "precision"
kfolds = 5

param_grid = [
              {'C':[1, 10, 100, 1000]}]


gs = SVC()
rand_search = RandomizedSearchCV(estimator = gs, param_distributions=param_grid, cv=kfolds, n_iter=300,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)
_ = rand_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

bestRecall= rand_search.best_estimator_

Fitting 5 folds for each of 4 candidates, totalling 20 fits
The best precision score is 0.6
... with parameters: {'C': 1}


C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\model_selection\_search.py:306: UserWarning: The total space of parameters 4 is smaller than n_iter=300. Running 4 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(
C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\utils\validation.py:1111: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


The precision value for SVM  rbf kernel with random search is 0.6 which is same as svm rbf kernel with grid search

Svm with poly kernel

In [58]:
score_measure = "precision"
kfolds = 5

param_grid = [
              {'C':[1, 10, 100, 1000]}]

PK = SVC()
grid_search = GridSearchCV(estimator =PK, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestRecallTree = grid_search.best_estimator_

Fitting 5 folds for each of 4 candidates, totalling 20 fits
The best precision score is 0.6
... with parameters: {'C': 1}


C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\utils\validation.py:1111: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


The precision score for SVM with poly kernel with grid search is 0.6.

In [59]:
score_measure = "precision"
kfolds = 5

param_grid = [
              {'C':[1, 10, 100, 1000]}]


PR = SVC()
rand_search = RandomizedSearchCV(estimator = PR, param_distributions=param_grid, cv=kfolds, n_iter=300,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)
_ = rand_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

bestRecall= rand_search.best_estimator_

Fitting 5 folds for each of 4 candidates, totalling 20 fits
The best precision score is 0.6
... with parameters: {'C': 1}


C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\model_selection\_search.py:306: UserWarning: The total space of parameters 4 is smaller than n_iter=300. Running 4 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(
C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\utils\validation.py:1111: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


The precision score for randomsearch with svm with poly kernel is 0.6 and as we can see we got the same precision value for all the SVM models 

fitting DTree

In [60]:
dtree1 = DecisionTreeClassifier().fit(X_train, np.ravel(y_train))

In [61]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier 
from matplotlib import pyplot as plt
import numpy as np
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV

In [62]:
score_measure = "precision"
kfolds = 5

param_grid = {
    'min_samples_split': np.arange(1,100),  
    'min_samples_leaf': np.arange(1,100),
    'min_impurity_decrease': np.arange(0.0001, 0.01, 0.0005),
    'max_leaf_nodes': np.arange(5, 100), 
    'max_depth': np.arange(1,50), 
    'criterion': ['entropy', 'gini'],
}


In [63]:
dtree = DecisionTreeClassifier()
rand_search = RandomizedSearchCV(estimator = dtree, param_distributions=param_grid, cv=kfolds, n_iter=300,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

In [64]:
_ = rand_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

bestRecallTree = rand_search.best_estimator_

Fitting 5 folds for each of 300 candidates, totalling 1500 fits
The best precision score is 0.5
... with parameters: {'min_samples_split': 6, 'min_samples_leaf': 2, 'min_impurity_decrease': 0.004600000000000001, 'max_leaf_nodes': 78, 'max_depth': 14, 'criterion': 'entropy'}


C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\model_selection\_validation.py:378: FitFailedWarning: 
25 fits failed out of a total of 1500.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
25 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\model_selection\_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\tree\_classes.py", line 969, in fit
    super().fit(
  File "C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\tree\_classes.py", line 265, in fit
    check_scalar(
  File "C:\

The precision score for Decision tree with random search is 0.53 which less compared to all other models.

In [65]:
score_measure = "precision"
kfolds = 5

param_grid = {
    'min_samples_split': np.arange(40,44),  
    'min_samples_leaf': np.arange(10,16),
    'min_impurity_decrease': np.arange(0.0001, 0.001,0.0005),
    'max_leaf_nodes': np.arange(60,70), 
    'max_depth': np.arange(15,20), 
    'criterion': ['entropy'],
}

dtree = DecisionTreeClassifier()
grid_search = GridSearchCV(estimator = dtree, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestRecallTree = grid_search.best_estimator_

Fitting 5 folds for each of 2400 candidates, totalling 12000 fits
The best precision score is 0.2333333333333333
... with parameters: {'criterion': 'entropy', 'max_depth': 15, 'max_leaf_nodes': 60, 'min_impurity_decrease': 0.0001, 'min_samples_leaf': 10, 'min_samples_split': 40}


The precision score for Decision tree with grid search is 0.23 which is way less than others.

Conclusion:

We can clearly see that the precision score for logistic regression is 0.7 and  which is higher compared to all other models for the given dataset.The best model fitting for this dataset is logistic model. So, the regeneration of lympography inside the body has a positive rate of 70 percent chances based on all other parameters.
    

In [66]:
from __future__ import print_function
import numpy as np
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, precision_score, recall_score
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
import matplotlib.pyplot as plt

from sklearn import datasets


np.random.seed(1)



In [67]:


ann = MLPClassifier(hidden_layer_sizes=(60,50,40), solver='adam', max_iter=200)
_ = ann.fit(X_train, y_train)

y_pred = ann.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           1       1.00      0.62      0.76        42
           2       0.16      1.00      0.27         3

    accuracy                           0.64        45
   macro avg       0.58      0.81      0.52        45
weighted avg       0.94      0.64      0.73        45



C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1118: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [68]:
#with random search cv


score_measure = "precision"
kfolds = 5

param_grid = {
    'hidden_layer_sizes': [ (50,), (70,),(50,30), (40,20), (60,40, 20), (70,50,40)],
    'activation': ['logistic', 'tanh', 'relu'],
    'solver': ['adam', 'sgd'],
    'alpha': [0, .2, .5, .7, 1],
    'learning_rate': ['constant', 'invscaling', 'adaptive'],
    'learning_rate_init': [0.001, 0.01, 0.1, 0.2, 0.5],
    'max_iter': [5000]
}

ann = MLPClassifier()
grid_search = RandomizedSearchCV(estimator = ann, param_distributions=param_grid, cv=kfolds, n_iter=100,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

bestRecallTree = grid_search.best_estimator_

print(grid_search.best_params_)

y_pred = bestRecallTree.predict(X_test)

print(classification_report(y_test, y_pred))

Fitting 5 folds for each of 100 candidates, totalling 500 fits
{'solver': 'adam', 'max_iter': 5000, 'learning_rate_init': 0.01, 'learning_rate': 'adaptive', 'hidden_layer_sizes': (40, 20), 'alpha': 0, 'activation': 'logistic'}
              precision    recall  f1-score   support

           1       1.00      0.60      0.75        42
           2       0.15      1.00      0.26         3

    accuracy                           0.62        45
   macro avg       0.57      0.80      0.50        45
weighted avg       0.94      0.62      0.71        45



C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1118: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [69]:


score_measure = "precision"
kfolds = 5

param_grid = {
    'hidden_layer_sizes': [ (30,), (50,), (70,), (90,)],
    'activation': ['tanh', 'relu'],
    'solver': ['adam'],
    'alpha': [.5, .7, 1],
    'learning_rate': ['adaptive', 'invscaling'],
    'learning_rate_init': [0.005, 0.01, 0.15],
    'max_iter': [5000]
}

ann = MLPClassifier()
grid_search = GridSearchCV(estimator = ann, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

bestRecallTree = grid_search.best_estimator_

print(grid_search.best_params_)
y_pred = bestRecallTree.predict(X_test)

print(classification_report(y_test, y_pred))

Fitting 5 folds for each of 144 candidates, totalling 720 fits
{'activation': 'tanh', 'alpha': 0.5, 'hidden_layer_sizes': (30,), 'learning_rate': 'adaptive', 'learning_rate_init': 0.005, 'max_iter': 5000, 'solver': 'adam'}
              precision    recall  f1-score   support

           1       1.00      0.67      0.80        42
           2       0.18      1.00      0.30         3

    accuracy                           0.69        45
   macro avg       0.59      0.83      0.55        45
weighted avg       0.95      0.69      0.77        45



C:\Users\ssand\AppData\Roaming\Python\Python39\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1118: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


After, adding neural network  to this dataset we can see the precision weight avg value with random search is 0.94 and precision weight avg value with grid search is 0.95. We got precison value for logistic regression is 0.7 clearly we can say the neural network  performmed better with this dataset and got an high precision value with both grid and random search for this dataset. If we compare in bewteen this two the grid search performmed better than random search overall in neural network.

